File: 498-diagonal-traverse/diagonal-traverse.js
var findDiagonalOrder = function(mat) {
    let rows = mat.length
    let columns = mat[0].length
    let goingUp = true
    let output = []
    let size = rows * columns
    let i = 0
    let j = 0

    while (output.length < size) {

        if (goingUp) {

            while (i >= 0 && j < columns) {
                output.push(mat[i][j])
                i -= 1
                j += 1
            }
            if (j == columns) {
                i += 2
                j -= 1
            }
            else i += 1
            goingUp = false

        }
        else {
            while (j >= 0 && i < rows) {
                output.push(mat[i][j])
                i += 1
                j -= 1
            }
            if (i == rows) {
                i -= 1
                j += 2
            }
            else j += 1
            goingUp = true
        }
    }
    return output
};

File: 498-diagonal-traverse/diagonal-traverse.py
class Solution:
    def findDiagonalOrder(self, mat: List[List[int]]) -> List[int]:
        going_up = True
        rows = len(mat)
        columns = len(mat[0])
        output = []
        i = 0
        j = 0

        while (len(output) < (rows * columns)):
            if going_up:
                while (i >= 0 and j < columns):
                    output.append(mat[i][j])
                    i -= 1
                    j += 1

                if (j == columns):
                    i += 2
                    j -= 1
                else:
                    i += 1
                going_up = False
            else:
                while (j >= 0 and i < rows):
                    output.append(mat[i][j])
                    j -= 1
                    i += 1

                if (i == rows):
                    j += 2
                    i -= 1
                else:
                    j += 1
                going_up = True


        return output


            